File: .perso/algo_matrice/DeepestDescent/C/grad.c
#include <stdio.h>
#include <stdlib.h>
#include "matrice.h"
#include <math.h>

void Exit_erreur(const char *message){
	printf("erreur : %s",message);
	exit(EXIT_FAILURE);
}

void main(){
	Tmatrice A,b,x,r,temp;
	int n;
	scanf("%d",&n);

/*allocation*/
	if(!Allouer(&A,n,n)) Exit_erreur("allocation");

	if(!Allouer(&b,n,1)) Exit_erreur("allocation");
	if(!Allouer(&x,n,1)) Exit_erreur("allocation"); 

	if(!Allouer(&r,n,1)) Exit_erreur("allocation");
     	if(!Allouer(&temp,n,1)) Exit_erreur("allocation");
     

/*remplissage*/

	if(!Remplir(&A)) Exit_erreur("remplissage");
	if(!Remplir(&b)) Exit_erreur("remplissage");
	if(!Remplir(&x)) Exit_erreur("remplissage");

/*main*/
	int Nmax;
	double tol,normeR,alpha;
	scanf("%d %lf",&Nmax,&tol);
	
	Matmul(A,x,&r);
	Mult(-1.,&r);
	Add(&r,b);
	for(int i=0;i<Nmax;i++){
		normeR = Dot_product(r,r);
		printf("%d %e\n",i,normeR);
		if(normeR < tol)
			 break;
		Matmul(A,r,&temp);
		alpha = normeR / Dot_product(temp,r);
		Mult(alpha,&r);
		Add(&x,r);
		Matmul(A,x,&r);
		Mult(-1.,&r);
		Add(&r,b);
	}
	printf("soltion :\n");
	Affiche(x);

/*desallocation*/
	if(!Desallouer(&A)) Exit_erreur("desallocation");
	if(!Desallouer(&b)) Exit_erreur("desallocation");
	if(!Desallouer(&x)) Exit_erreur("desallocation");
	if(!Desallouer(&r)) Exit_erreur("desallocation");
	if(!Desallouer(&temp)) Exit_erreur("desallocation");
	
}

File: .perso/C/algo_matrice/DeepestDescent/Fortran/grad.f
        program grad              
        implicit none
        double precision,dimension(:,:),allocatable :: A
        double precision,dimension(:), allocatable :: b,X,r
        integer n,i,nMax
        double precision tol,normeR,alpha

        read(*,*) n,nMax,tol

        allocate(A(n,n))
        allocate(b(n))
        allocate(X(n))
        allocate(r(n))

        do i=1,n
                read(*,*) A(i,1:n)
        enddo
        read(*,*)
        read(*,*) b
        read(*,*) X

        r = b - matmul(A,x)

        do i =1,nMax
                normeR = dot_product(r,r)
                write(*,*) i , normeR
                if(normeR<tol) then
                        exit
                endif
                alpha = normeR/dot_product(matmul(A,r),r)
                X = X + alpha * r
                r = b - matmul(A,x)        
        enddo        


        
       
        write(*,*) X
       

        deallocate(A)
        deallocate(b)
        deallocate(X)
        deallocate(r)
        end
                
        
